hal/src/boron/network/sara_ncp_client.cpp

```diff
@@ -890,32 +924,37 @@ int SaraNcpClient::initReady() {
     if (conf_.ncpIdentifier() != PLATFORM_NCP_SARA_R410) {
         // Change the baudrate to 921600
         CHECK(changeBaudRate(UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_U2));
-        // Check that the modem is responsive at the new baudrate
-        skipAll(serial_.get(), 1000);
-        CHECK(waitAtResponse(10000));
-    }
+    } else {
+        int fwVersion = getAppFirmwareVersion();
+        if (fwVersion > 0) {
+            // L0.0.00.00.05.06,A.02.00 has a memory issue
+            memoryIssuePresent_ = (fwVersion == UBLOX_NCP_R4_APP_FW_VERSION_MEMORY_LEAK_ISSUE);
+            // There is a set of other revisions which do not have hardware flow control
+            if (!(fwVersion >= UBLOX_NCP_R4_APP_FW_VERSION_NO_HW_FLOW_CONTROL_MIN &&
+                    fwVersion <= UBLOX_NCP_R4_APP_FW_VERSION_NO_HW_FLOW_CONTROL_MAX)) {
+                // Change the baudrate to 460800
+                // NOTE: ignoring AT errors just in case to accommodate for some revisions
+                // potentially not supporting anything other than 115200
+                int r = changeBaudRate(UBLOX_NCP_RUNTIME_SERIAL_BAUDRATE_R4);
+                if (r != SYSTEM_ERROR_AT_NOT_OK) {
+                    return r;
+                }
 
-    if (ncpId() == PLATFORM_NCP_SARA_R410) {
-        // ATI9 (get version and app version)
-        // example output
-        // 16 "\r\n08.90,A01.13\r\n" G350 (newer)
-        // 16 "\r\n08.70,A00.02\r\n" G350 (older)
-        // 28 "\r\nL0.0.00.00.05.06,A.02.00\r\n" (memory issue)
-        // 28 "\r\nL0.0.00.00.05.07,A.02.02\r\n" (demonstrator)
-        // 28 "\r\nL0.0.00.00.05.08,A.02.04\r\n" (maintenance)
-        auto resp = parser_.sendCommand("ATI9");
-        char verExtended[33] = {};
-        memoryIssuePresent_ = false;
-        int r = CHECK_PARSER(resp.scanf("%32[^\n]", verExtended));
-        CHECK_TRUE(r == 1, SYSTEM_ERROR_AT_RESPONSE_UNEXPECTED);
-        r = CHECK_PARSER(resp.readResult());
-        CHECK_TRUE(r == AtResponse::OK, SYSTEM_ERROR_AT_NOT_OK);
-        if (!strcmp(verExtended, "L0.0.00.00.05.06,A.02.00")) {
-            memoryIssuePresent_ = true;
+                if (r == SYSTEM_ERROR_NONE) {
+                    // Make sure flow control is enabled as well
+                    CHECK_PARSER_OK(parser_.execCommand("AT+IFC=2,2"));
+                }
```

Comment:
This won't be executed unless the above call to `changeBaudRate()` fails with `SYSTEM_ERROR_AT_NOT_OK`

Reply:
The condition is `if (r == SYSTEM_ERROR_NONE) {` though, so it only gets executed if the above call to change the baud rate succeeds, right?

Reply:
The above call returns from the function on any value including SYSTEM_ERROR_NONE

Reply:
Oh, right. :+1: Let's fix that.